codebuild/bin/s2n_dynamic_load_test.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <dlfcn.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+static void *s2n_load_dynamic_lib(void *ctx)
+{
+    const char *s2n_so_path = ctx;
+
+    void *s2n_so = dlopen(s2n_so_path, RTLD_NOW);
+    if (!s2n_so) {
+        exit(1);
+    }
+
+    int (*s2n_init_dl)(void) = NULL;
+    *(void **) (&s2n_init_dl) = dlsym(s2n_so, "s2n_init");
+    if (dlerror()) {
+        exit(1);
+    }
+
+    int (*s2n_cleanup_dl)(void) = NULL;
+    *(void **) (&s2n_cleanup_dl) = dlsym(s2n_so, "s2n_cleanup");
+    if (dlerror()) {
+        exit(1);
+    }
+
+    if ((*s2n_init_dl)()) {
+        exit(1);
+    }
+    if ((*s2n_cleanup_dl)()) {
+        exit(1);
+    }
```

Comment:
Can we also add a test case scenario where we dont call `cleanup` and verify that this doesnt break

Reply:
If you don't call cleanup this will blow up because the dangling destructor will get invoked after unload.  If you want a solution where that doesn't happen then you'd need to completely rethink how thread locals and cleanup are managed.

Reply:
The dlclose call will technically calls s2n_cleanup because of our atexit handler. So there wouldn't be a dangling destructor in this case.

> https://linux.die.net/man/3/dlopen
Since glibc 2.2.3, atexit can be used to register an exit handler that is automatically called when a library is unloaded.

Although it is dependent on the glibc version I guess.

Reply:
Right I expect this to not fail if an application fails to all s2n_cleanup since the atexit callback will be invoked. Do we want to test that case here as well to confirm that